Adds rapidcheck gtest/catch integrations

mjvankampen opened the pull request:
Specify library name and version:  **rapidcheck**

Rapidcheck provides integration with multiple unit test frameworks. This enables those integrations within the conan package

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

mjvankampen commented:
Excuse the mess, not my finest moment

prince-chrismc commented on recipes/rapidcheck/all/conanfile.py:
```diff
@@ -103,15 +123,33 @@ def _module_file_rel_path(self):
-                self.cpp_info.defines.append("RC_DONT_USE_RTTI")
+                self.cpp_info.components["core"].defines.append("RC_DONT_USE_RTTI")
+                
+        if(self.options.enable_catch):
+            self.cpp_info.components["catch"].requires = ["core"]
+            self.cpp_info.components["catch"].includedirs  = ["include"]
+        if(self.options.enable_gmock):
+            self.cpp_info.components["gmock"].requires = ["core"]
+            self.cpp_info.components["gmock"].includedirs  = ["include"]
+        if(self.options.enable_gtest):
+            self.cpp_info.components["gtest"].requires = ["core"]
+            self.cpp_info.components["gtest"].includedirs  = ["include"]
+        if(self.options.enable_boost):
+            self.cpp_info.components["boost"].requires = ["core"]
+            self.cpp_info.components["boost"].includedirs  = ["include"]
+        if(self.options.enable_boost_test):
+            self.cpp_info.components["boost_test"].requires = ["core"]
+            self.cpp_info.components["boost_test"].includedirs  = ["include"]
```
include is the default? https://docs.conan.io/en/latest/reference/conanfile/attributes.html#cpp-info

<html>
<body>
<!--StartFragment-->

NAME | DESCRIPTION
-- | --
self.cpp_info.includedirs | Ordered list with include paths. Defaulted to ["include"]

<!--EndFragment-->
</body>
</html>

```suggestion
        if(self.options.enable_catch):
            self.cpp_info.components["catch"].requires = ["core"]
        if(self.options.enable_gmock):
            self.cpp_info.components["gmock"].requires = ["core"]
        if(self.options.enable_gtest):
            self.cpp_info.components["gtest"].requires = ["core"]
        if(self.options.enable_boost):
            self.cpp_info.components["boost"].requires = ["core"]
        if(self.options.enable_boost_test):
            self.cpp_info.components["boost_test"].requires = ["core"]
```

prince-chrismc commented on recipes/rapidcheck/all/conanfile.py:
```diff
@@ -20,11 +20,21 @@ class RapidcheckConan(ConanFile):
         "shared": [True, False],
         "fPIC": [True, False],
         "enable_rtti": [True, False],
+        "enable_catch": [True, False],
+        "enable_gmock": [True, False],
+        "enable_gtest": [True, False],
+        "enable_boost": [True, False],
+        "enable_boost_test": [True, False]
```
Could this be one option?

```suggestion
        "enable_framework": [False, "catch", "gmock", "gtest", "boost", "boost_test"],
```

mjvankampen replied:
No I don't think so as you can enable multiple integrations at the same time.

prince-chrismc reviewed (commented):
The component names do not match the official ones 😕 

https://github.com/emil-e/rapidcheck/search?q=rapidcheckConfig

- core --> rapidcheck
- the frameworks should be prefixed with `rapidcheck_`

***

Also why are there no require statements? I would expect "enable_catch" to include catch..
https://github.com/emil-e/rapidcheck/blob/fc616114a95a3198514f8b4ec9e354865fd6ecfc/extras/catch/include/rapidcheck/catch.h#L6

Would fail to compile without it

mjvankampen commented:
Requires: I see, catch is handled differently from gtest. Gtest requires the consumer to link gtest, include gtest header and then include rapidcheck header which makes sense to not require gtest. Different for catch. Will check it out.

Names: I think the official targets are not namespaced correct? This is why the aliases with the correct names are added without a namespace.

Core is just the name of the component, it cannot be the same name as the package I think. The resulting target is still named rapidcheck::rapidcheck (and aliased to rapidcheck)

mjvankampen commented:
Please have a look. I left out boost for now. I think it is not very nice to link gtest/catch to the core lib, but I would like to keep the gtest and such integration libs INTERFACE libraries to not change the code around too much.

prince-chrismc commented:
> I think it is not very nice to link gtest/catch to the core lib, but I would like to keep the gtest and such integration libs INTERFACE libraries to not change the code around too much.

I absolutely agree, its completely customer facing.
You should have added it to the framework components not to "core" 

I am not sure why you added patches since that was not required in the first place 🤔 

The CMake aliases are perfect, my previous note was about the Conan ones... when you go to use them in a downstream conanfile the sames are not very distinctive (they are the names of other projects) without the prefix that upstream uses

mjvankampen commented:
Please have a look if this addresses your concerns:
1. I added the external libs to core as they are introduced to into the upstream project as cmake subdirectory. I don't think we can do anything similar on conan so I thought this was closest. I agree it is not.
2. A patch is required to disable the git submodules. Agreed that the rest of the patch is not required.
3. Clear updated

prince-chrismc reviewed (approved):
Amazing work!

uilianries commented:
@prince-chrismc Good catch! thanks for reviewing!

uilianries commented on recipes/rapidcheck/all/conanfile.py:
```diff
@@ -62,10 +74,15 @@ def _configure_cmake(self):
         self._cmake.configure(build_folder=self._build_subfolder)
         return self._cmake
 
     def build(self):
```
```suggestion
    def build(self):
        if self.options.enable_gmock and not self.deps_cpp_info["gtest"].build_gmock:
            raise ConanInvalidConfiguration("The option `rapidcheck:enable_gmock` requires gtest:build_gmock=True`")
```

The gmock is optional for Gtest. By default it's enabled, but still anyone can use it disabled. It should be checked in build(), because deps_cpp_info will be populated only before build method.

uilianries reviewed (approved):
LGTM
